src/main/java/com/sarality/action/ActionChain.java

```diff
@@ -12,25 +14,37 @@
 public class ActionChain implements ViewAction {
 
   private final List<ViewAction> actionList = new ArrayList<>();
+  private final ActionChain failureActionChain;
 
   public ActionChain(ViewAction... actions) {
-    if (actions == null || actions.length < 1) {
+    this((actions == null ? new ArrayList<ViewAction>() : Arrays.asList(actions)));
+  }
+
+  public ActionChain(List<ViewAction> actions, ViewAction... failureActions) {
+    if (actions == null || actions.size() < 1) {
       throw new IllegalArgumentException("Must specify at least the initial action for an action chain");
     }
-    this.actionList.addAll(Arrays.asList(actions));
+    this.actionList.addAll(actions);
+
+    if (failureActions != null && failureActions.length > 0) {
+      this.failureActionChain = new ActionChain(failureActions);
+    } else {
+      this.failureActionChain = null;
+    }
+
   }
-  
+
   @Override
   public boolean perform(ActionContext actionContext) {
     boolean success = true;
     for (ViewAction action : actionList) {
-      // TODO(abhideep): Add support for onFailure based chaining.
       success = action.perform(actionContext);
       if (!success) {
         break;
       }
     }
 
```

Comment:
Might make sense to call out failureAction explicitly

if (!success && failureAction != null) {
  return failureAction.perform(actionContext);
}

return success;